src/main/java/com/redhat/rhjmc/containerjfr/net/web/WebServer.java

```diff
@@ -0,0 +1,327 @@
+        }
+
+        return response;
+    }
+
+    private void endWithReport(InputStream recording, String recordingName, HttpServerResponse response) throws IOException, CouldNotLoadRecordingException {
+        // blocking function, must be called from a blocking handler
+        try (recording) {
+            response.end(reportGenerator.generateReport(recording));
+
+            // ugly hack for "trimming" created clones of specified recording. JMC service creates a clone of running
+            // recordings before loading events to create the report, and these clones are erroneously left dangling.
+            TRIM_WORKER.submit(() -> {
+                try {
+                    service.getAvailableRecordings()
+                            .stream()
+                            .filter(r -> r.getName().equals(String.format("Clone of %s", recordingName)))
+                            .forEach(r -> {
+                                try {
+                                    service.close(r);
+                                } catch (FlightRecorderException fre) {
+                                    logger.debug(fre);
+                                }
+                            });
+                } catch (FlightRecorderException fre) {
+                    logger.debug(fre);
+                }
+            });
+        }
+    }
+
+    void handleClientUrlRequest(RoutingContext ctx) {
+        ctx.response().putHeader(HttpHeaders.CONTENT_TYPE, MIME_TYPE_JSON);
+        try {
+            endWithJsonKeyValue("clientUrl", String.format("ws://%s:%d/command", netConf.getWebServerHost(), netConf.getExternalWebServerPort()), ctx.response());
+        } catch (SocketException | UnknownHostException e) {
+            throw new HttpStatusException(500, e);
+        }
+    }
+
+    void handleGrafanaDatasourceUrlRequest(RoutingContext ctx) {
+        ctx.response().putHeader(HttpHeaders.CONTENT_TYPE, MIME_TYPE_JSON);
+        endWithJsonKeyValue("grafanaDatasourceUrl", env.getEnv(GRAFANA_DATASOURCE_ENV, ""), ctx.response());
+    }
+
+    void handleGrafanaDashboardUrlRequest(RoutingContext ctx) {
+        ctx.response().putHeader(HttpHeaders.CONTENT_TYPE, MIME_TYPE_JSON);
+        endWithJsonKeyValue("grafanaDashboardUrl", env.getEnv(GRAFANA_DASHBOARD_ENV, ""), ctx.response());
+    }
+
+    void handleRecordingDownloadRequest(String recordingName, RoutingContext ctx) {
+        try {
+            Optional<InputStream> recording = getRecordingInputStream(recordingName);
+            if (recording.isEmpty()) {
+                throw new HttpStatusException(404, String.format("%s not found", recordingName));
+            }
+
+            ctx.response().setChunked(true);
+            ctx.response().putHeader(HttpHeaders.CONTENT_TYPE, MIME_TYPE_OCTET_STREAM);
+            writeInputStream(recording.get(), ctx.response())
+                    .endHandler((e) -> downloadCounts.merge(recordingName, 1, Integer::sum))
```

Comment:
Fancy. :-)

Reply:
(This isn't a complaint, I just like the `merge`)

Reply:
My IDE promoted this suggested change. Very fancy indeed. XD